pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -0,0 +1,867 @@
+					Reason:             "ListenerReady",
+					Message:            "No error found",
+				})
+			} else {
+				listenerInError = true
+			}
+		}
+
+		gatewayStatus.Listeners = listenerStatuses
+
+		if listenerInError {
+			// GatewayConditionReady "Ready", GatewayConditionReason "ListenersNotValid"
+			gatewayStatus.Conditions = append(gatewayStatus.Conditions, metav1.Condition{
+				Type:               string(v1alpha1.GatewayConditionReady),
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: metav1.Now(),
+				Reason:             string(v1alpha1.GatewayReasonListenersNotValid),
+				Message:            "All Listeners must be valid",
+			})
+
+			err := client.UpdateGatewayStatus(gateway, gatewayStatus)
+			if err != nil {
+				logger.Debugf("An error occurred while updating gateway status: %v", err)
+			}
+
+			continue
+		}
+
+		// update "Scheduled" status with "ResourcesAvailable" reason
+		gatewayStatus.Conditions = append(gatewayStatus.Conditions, metav1.Condition{
+			Type:               string(v1alpha1.GatewayConditionScheduled),
+			Status:             metav1.ConditionTrue,
+			LastTransitionTime: metav1.Now(),
+			Reason:             "ResourcesAvailable",
+			Message:            "Found matching HTTPRoutes",
+		})
+
+		// update "Ready" status with "ListenersValid" reason
+		gatewayStatus.Conditions = append(gatewayStatus.Conditions, metav1.Condition{
+			Type:               string(v1alpha1.GatewayConditionReady),
+			Status:             metav1.ConditionTrue,
+			LastTransitionTime: metav1.Now(),
+			Reason:             "ListenersValid",
+			Message:            "All listeners are valid",
+		})
+
+		err := client.UpdateGatewayStatus(gateway, gatewayStatus)
+		if err != nil {
+			logger.Debugf("An error occurred while updating gateway status: %v", err)
+		}
+
+		if len(routers) > 0 {
+			for k, router := range routers {
+				// TODO manage duplicate key
+				conf.HTTP.Routers[k] = router
+			}
+		}
+
+		if len(services) > 0 {
+			for k, service := range services {
+				// TODO manage duplicate key
+				conf.HTTP.Services[k] = service
+			}
+		}
+
+		if len(tlsConfigs) > 0 {
+			if conf.TLS == nil {
+				conf.TLS = &dynamic.TLSConfiguration{}
+			}
+			conf.TLS.Certificates = append(conf.TLS.Certificates, getTLSConfig(tlsConfigs)...)
+		}
+	}
+
+	return conf
+}
+
+func hostRule(httpRouteSpec v1alpha1.HTTPRouteSpec) string {
+	if len(httpRouteSpec.Hostnames) > 0 {
+		hostRule := ""
+		for i, hostname := range httpRouteSpec.Hostnames {
+			if i > 0 && len(hostname) > 0 {
+				hostRule += "`, `"
+			}
+			hostRule += string(hostname)
+		}
+
+		if hostRule != "" {
+			return "Host(`" + hostRule + "`)"
+		}
+	}
+
+	return ""
+}
+
+func extractRule(routeRule v1alpha1.HTTPRouteRule, hostRule string) string {
+	var rule string
+	var matchesRules []string
+
+	for _, match := range routeRule.Matches {
+		var matchRules []string
+		// TODO handle other path types
+		if match.Path.Type == v1alpha1.PathMatchExact {
```

Comment:
Maybe it's just a matter of taste but I would have used a `switch` here, and add a log if the user has used an unsupported path `Type` matching, wdyt?